diff --git a/node_modules/react-use-websocket/dist/lib/use-websocket.js b/node_modules/react-use-websocket/dist/lib/use-websocket.js
index f01db48..b30aff2 100644
--- a/node_modules/react-use-websocket/dist/lib/use-websocket.js
+++ b/node_modules/react-use-websocket/dist/lib/use-websocket.js
@@ -139,15 +139,15 @@ var useWebSocket = function (url, options, connect) {
                             }
                             protectedSetLastMessage = function (message) {
                                 if (!expectClose_1) {
-                                    (0, react_dom_1.flushSync)(function () { return setLastMessage(message); });
+                                    setLastMessage(message);
                                 }
                             };
                             protectedSetReadyState = function (state) {
                                 if (!expectClose_1) {
-                                    (0, react_dom_1.flushSync)(function () { return setReadyState(function (prev) {
+                                    setReadyState(function (prev) {
                                         var _a;
                                         return (__assign(__assign({}, prev), (convertedUrl.current && (_a = {}, _a[convertedUrl.current] = state, _a))));
-                                    }); });
+                                    });
                                 }
                             };
                             if (createOrJoin_1) {
