// framer-motion@11.18.2/dist/es/dom.mjs
export { isDragActive } from 'motion-dom';
export { invariant, noop, progress } from 'motion-utils';
export { animate, createScopedAnimate } from './animation/animate/index.mjs';
export { animateMini } from './animation/animators/waapi/animate-style.mjs';
export { scroll } from './render/dom/scroll/index.mjs';
export { scrollInfo } from './render/dom/scroll/track.mjs';
export { inView } from './render/dom/viewport/index.mjs';
export { MotionValue, motionValue } from './value/index.mjs';
export { anticipate } from './easing/anticipate.mjs';
export { backIn, backInOut, backOut } from './easing/back.mjs';
export { circIn, circInOut, circOut } from './easing/circ.mjs';
export { cubicBezier } from './easing/cubic-bezier.mjs';
export { easeIn, easeInOut, easeOut } from './easing/ease.mjs';
export { mirrorEasing } from './easing/modifiers/mirror.mjs';
export { reverseEasing } from './easing/modifiers/reverse.mjs';
export { steps } from './easing/steps.mjs';
export { inertia } from './animation/generators/inertia.mjs';
export { keyframes } from './animation/generators/keyframes.mjs';
export { spring } from './animation/generators/spring/index.mjs';
export { stagger } from './animation/utils/stagger.mjs';
export { time } from './frameloop/sync-time.mjs';
export { clamp } from './utils/clamp.mjs';
export { delayInSeconds as delay } from './utils/delay.mjs';
export { distance, distance2D } from './utils/distance.mjs';
export { interpolate } from './utils/interpolate.mjs';
export { mix } from './utils/mix/index.mjs';
export { pipe } from './utils/pipe.mjs';
export { transform } from './utils/transform.mjs';
export { wrap } from './utils/wrap.mjs';
export { cancelSync, sync } from './frameloop/index-legacy.mjs';
export { cancelFrame, frame, frameData, frameSteps } from './frameloop/frame.mjs';
