// monaco-editor@0.53.0/dev/vs/javascript.00c827f5.js
define("vs/javascript.00c827f5", ["exports", "./typescript.21e6b24c", "./editor.api.001a2486"], function(exports, typescript, editor_api) {
  "use strict";
  const conf = typescript.conf;
  const language = {
    defaultToken: "invalid",
    tokenPostfix: ".js",
    keywords: [
      "break",
      "case",
      "catch",
      "class",
      "continue",
      "const",
      "constructor",
      "debugger",
      "default",
      "delete",
      "do",
      "else",
      "export",
      "extends",
      "false",
      "finally",
      "for",
      "from",
      "function",
      "get",
      "if",
      "import",
      "in",
      "instanceof",
      "let",
      "new",
      "null",
      "return",
      "set",
      "static",
      "super",
      "switch",
      "symbol",
      "this",
      "throw",
      "true",
      "try",
      "typeof",
      "undefined",
      "var",
      "void",
      "while",
      "with",
      "yield",
      "async",
      "await",
      "of"
    ],
    typeKeywords: [],
    operators: typescript.language.operators,
    symbols: typescript.language.symbols,
    escapes: typescript.language.escapes,
    digits: typescript.language.digits,
    octaldigits: typescript.language.octaldigits,
    binarydigits: typescript.language.binarydigits,
    hexdigits: typescript.language.hexdigits,
    regexpctl: typescript.language.regexpctl,
    regexpesc: typescript.language.regexpesc,
    tokenizer: typescript.language.tokenizer
  };
  exports.conf = conf;
  exports.language = language;
  Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
});
